File: Hackerrank/Medium/check_BST/check_BST.java
/* Hidden stub code will pass a root argument to the function below. Complete the function to solve the challenge. Hint: you may want to write one or more helper functions.  

The Node class is defined as follows:
    class Node {
        int data;
        Node left;
        Node right;
     }
*/
    //we will be storing the temp. node to see if prev node is less than right node.
    Node prev;
    boolean checkBST(Node root) {
        //if root is null return true
        if (root == null)
            return (true);
        
        //using in order traversal to check BST
        //is left node is null??
        if (!checkBST(root.left))
        {
            return (false);
        }

        //if there stored previous node and prev.data is GREATER than root.data then return FALSE
        if (prev != null && prev.data >= root.data)
        {
            return (false);
        }
        //store the current root node
        prev = root;
        //go to the right node...
        return (checkBST(root.right));
            
        
        
    }

File: Python/Practices/Trees/check_bst.py
'''
given a binary tree, check if it is BST or not

        6
    3      8
1     4  5   9

return True or False
traverse(head,ancestor)
    if head == None
        return False
    if not(head.right) and head.left:
        return traverse(head.right)
    if not(head.left) and head.right:
        return traverse(head.left)
    if head.left.data >= head.data and head.right.data <= head.data
        return False
    if head.left.data < head.data and head.right.data > head.data
        if ancestor and ancestor.data < head.left.data and ancestor.data > head.r
            return traverse(head.left,head) and traverse(head.right,head)

'''
